--- a/ggml/src/ggml-cpu/arch/arm/quants.c
+++ b/ggml/src/ggml-cpu/arch/arm/quants.c
@@ -2821,8 +2821,8 @@ void ggml_vec_dot_q4_K_q8_K(int n, float * GGML_RESTRICT s, size_t bs, const voi
         const uint8_t * GGML_RESTRICT q4 = x[i].qs;
         const int8_t  * GGML_RESTRICT q8 = y[i].qs;
 
-        int32_t sumi1 = 0;
-        int32_t sumi2 = 0;
+        int32x4_t sumi1 = vdupq_n_s32(0);
+        int32x4_t sumi2 = vdupq_n_s32(0);
 
         for (int j = 0; j < QK_K/64; ++j) {
             const ggml_uint8x16x2_t q4bits = ggml_vld1q_u8_x2(q4); q4 += 32;
@@ -2832,7 +2832,7 @@ void ggml_vec_dot_q4_K_q8_K(int n, float * GGML_RESTRICT s, size_t bs, const voi
             q4bytes.val[1] = vreinterpretq_s8_u8(vandq_u8  (q4bits.val[1], m4b));
 
             const int32x4_t p1 = ggml_vdotq_s32(ggml_vdotq_s32(mzero, q4bytes.val[0], q8bytes.val[0]), q4bytes.val[1], q8bytes.val[1]);
-            sumi1 += vaddvq_s32(p1) * scales[2*j+0];
+            sumi1 = vmlaq_n_s32(sumi1, p1, scales[2*j+0]);
 
             q8bytes = ggml_vld1q_s8_x2(q8); q8 += 32;
             q4bytes.val[0] = vreinterpretq_s8_u8(vshrq_n_u8(q4bits.val[0], 4));
@@ -2840,9 +2840,9 @@ void ggml_vec_dot_q4_K_q8_K(int n, float * GGML_RESTRICT s, size_t bs, const voi
 
             const int32x4_t p2 = ggml_vdotq_s32(ggml_vdotq_s32(mzero, q4bytes.val[0], q8bytes.val[0]), q4bytes.val[1], q8bytes.val[1]);
 
-            sumi2 += vaddvq_s32(p2) * scales[2*j+1];
+            sumi2 = vmlaq_n_s32(sumi2, p2, scales[2*j+1]);
         }
 
-        sumf += d * (sumi1 + sumi2);
+        sumf += d * vaddvq_s32(vaddq_s32(sumi1, sumi2));
 
     }
